xapian-core/tests/unittest.cc

```diff
@@ -536,10 +536,22 @@ static bool test_tostring1()
     TEST_EQUAL(str(-1), "-1");
     TEST_EQUAL(str(-9), "-9");
     TEST_EQUAL(str(-10), "-10");
+    TEST_EQUAL(str(0x7f), "127");
+    TEST_EQUAL(str(-0x80), "-128");
+    TEST_EQUAL(str(0x7fff), "32767");
+    // test for lvalue with minimum value for short type.
+    short temp = -32768;
+    TEST_EQUAL(str(temp), "-32768");
```

Comment:
This isn't necessarily the minimum value for `short` - the minimum must be <= -32767, but `short` could be wider (an implementation where sizeof(short) > 2 is allowed by the standard, and there have been such implementations).  All modern systems seem to use twos' complement, but on anything that doesn't -32768 may not be representable.

The minimum value for `short` is available using `std::numeric_limits<short>::min()` (which is also self-documenting).  Similarly for other standard types.

Reply:
oh you mean the comment?.

Reply:
The minimum value that a `short` can represent is a compile-time constant, but the value of that constant is not the same for every compiler.  And somewhat unhelpfully for us here the standard allows an implementation where `-32768` can't be represented as a `short`.  In practice I suspect there may already be some twos' complement assumptions in Xapian, but I feel it's a bad idea to knowingly add more such assumptions.

Taking a step back, fundamentally what we want to be adding testing for here is the edge cases where the input value is the maximum and minimum values of each integer type.  The `parseunsigned1` testcase I added recently does that quite nicely for `parse_unsigned()` - we just instantiate a template function for each type we want to test and then call it.

If we try to follow that model here, the only tricky bit seems to be generating the expected value to test against - we can't write out the expected answers like we currently do, and converting the result of `str()` back to an integer means we're testing each function in terms of the other and also we wouldn't spot some potential bugs - e.g. if `str(0)` -> `00` that would convert back to integer `0`.  But we can generate the expected strings with e.g. `std::ostringstream` - that's best avoided in production code as it is rather slow, but here can serve as our "ground truth" for converting integers to strings.

Reply:
thanks will change it.